packages/template-retail-react-app/app/hooks/use-datacloud.js

```diff
@@ -0,0 +1,481 @@
+        // 3. masterId - SKU of the Master Product
+        return {
+            id: product?.id ?? product?.productId ?? product?.master?.masterId
+        }
+    }
+    /**
+     * Constructs the base search result object with relevant search
+     * metadata.
+     *
+     * @param {object} searchParams - The searchParams object
+     * @returns {object} - The base search result object
+     */
+    _constructBaseSearchResult(searchParams) {
+        return {
+            searchResultTitle: searchParams.q,
+            searchResultPosition: searchParams.offset,
+            searchResultPageNumber:
+                searchParams.limit != 0 ? searchParams.offset / searchParams.limit + 1 : 1
+        }
+    }
+
+    _concatenateEvents = (...events) => ({...events.reduce((acc, obj) => ({...acc, ...obj}), {})})
+
+    /**
+     * Sends a `page-view` event to Data Cloud.
+     *
+     * This method records an `userEnagement` event type to track which page the shopper has viewed.
+     *
+     * @param {string} path - The URL path of the page that was viewed
+     * @param {object} args - Additional metadata for the event
+     */
+    async sendViewPage(path, args) {
+        const baseEvent = this._constructBaseEvent(args)
+
+        const identityProfile = this._concatenateEvents(
+            baseEvent,
+            this._generateEventDetails('identity', 'Profile'),
+            {
+                isAnonymous: args.isGuest,
+                firstName: args.firstName,
+                lastName: args.lastName,
+                sourceUrl: path
+            }
+        )
+
+        const userEngagement = this._concatenateEvents(
+            baseEvent,
+            this._generateEventDetails('userEngagement', 'Engagement'),
+            {
+                interactionName: 'page-view',
+                sourceUrl: path
+            }
+        )
+
+        const interaction = {
+            events: [identityProfile, userEngagement]
+        }
+
+        try {
+            this.sdk.webEventsAppSourceIdPost(interaction)
+        } catch (err) {
+            logger.error('Error sending DataCloud event')
+        }
+    }
+
+    /**
+     * Sends a `catalog-object-view-start` event to Data Cloud.
+     *
+     * This method records a `catalog` event type to track when a shopper
+     * views the details of a product (e.g. a Product Detail Page).
+     *
+     * @param {object} product - The product being viewed
+     * @param {object} args - Additional metadata for the event
+     */
+    async sendViewProduct(product, args) {
+        const baseEvent = this._constructBaseEvent(args)
+        const baseProduct = this._constructDatacloudProduct(product)
+
+        const identityProfile = this._concatenateEvents(
+            baseEvent,
+            this._generateEventDetails('identity', 'Profile'),
+            {
+                isAnonymous: args.isGuest,
+                firstName: args.firstName,
+                lastName: args.lastName
+            }
```

Comment:
Can we create a function to return user profile instead of having to passing args on `_concatenateEvents` on each method? 🤔